File: src/main/java/club/sk1er/hytilities/handlers/chat/ChatSendModule.java
package club.sk1er.hytilities.handlers.chat;

import net.minecraftforge.client.event.ClientChatReceivedEvent;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Since Forge lacks a {@link ClientChatReceivedEvent} for <em>sending</em> messages, this interface
 * is used to implement such an event. Register your class in {@link ChatHandler#ChatHandler()} and
 * it will be executed whenever the user sends a chat message.
 *
 * @see ChatModule
 * @see ChatHandler
 */
public interface ChatSendModule extends ChatModule {

    /**
     * Place your code here. Called when the user sends a chat message.
     *
     * @param message message that the user sent
     * @return the optionally modified message, or {@code null} if the message is to be cancelled
     */
    @Nullable
    String onMessageSend(@NotNull String message);

}

File: src/main/kotlin/org/polyfrost/hytils/client/handlers/chat/ChatSendModule.kt
package org.polyfrost.hytils.client.handlers.chat

import org.polyfrost.hytils.client.events.ChatSendEvent
import org.polyfrost.oneconfig.api.event.v1.EventManager

/**
 * This interface is used to handle many different [ChatSendEvent]-consuming methods
 * without having to add them all to the [EventManager].
 *
 * Must be registered in [ChatHandler] to be executed.
 *
 * @see ChatModule
 * @see ChatHandler
 */
interface ChatSendModule : ChatModule {
    /**
     * Place your code here. Called when a Hytils [ChatSendEvent] is received.
     *
     * If the event is canceled, [ChatModule]s after that event will not execute. Therefore,
     * [ChatSendEvent.cancelled]} checks are unnecessary.
     *
     * @param event a [ChatSendEvent]
     */
    fun onChatSend(event: ChatSendEvent)
}
